tools/swift-plugin-server/Sources/swift-plugin-server/WasmEngine.swift

```diff
@@ -10,42 +10,71 @@
 //
 //===----------------------------------------------------------------------===//
 
+#if canImport(Darwin)
+import Darwin
+#elseif canImport(Glibc)
+import Glibc
+#elseif canImport(Musl)
+import Musl
+#elseif canImport(WASILibc)
+import WASILibc
+#else
+#error("Unsupported platform")
+#endif
+
 import SystemPackage
 import WASI
 import WasmTypes
 
 typealias WasmFunction = () throws -> Void
 
 protocol WasmEngine {
-  init(path: FilePath, imports: WASIBridgeToHost) throws
+  init(path: FilePath, stdinPath: FilePath, stdoutPath: FilePath) throws
 
   func function(named name: String) throws -> WasmFunction?
 
+  func writeToPlugin(_ storage: some Sequence<UInt8>) throws
+  func readFromPlugin(into storage: UnsafeMutableRawBufferPointer) throws -> Int
+
   func shutDown() throws
 }
 
 typealias DefaultWasmPlugin = WasmEnginePlugin<DefaultWasmEngine>
 
 // a WasmPlugin implementation that delegates to a WasmEngine
 struct WasmEnginePlugin<Engine: WasmEngine>: WasmPlugin {
-  private let hostToPlugin: FileDescriptor
-  private let pluginToHost: FileDescriptor
+  enum Error: Swift.Error {
+    case failedToCreateNamedPipe(FilePath)
+  }
+
   private let pumpFunction: WasmFunction
+  private let tempDirectory: FilePath
+  private let stdinPath: FilePath
+  private let stdoutPath: FilePath
   let engine: Engine
 
   init(path: FilePath) throws {
-    let hostToPluginPipes = try FileDescriptor.pipe()
-    let pluginToHostPipes = try FileDescriptor.pipe()
-    self.hostToPlugin = hostToPluginPipes.writeEnd
-    self.pluginToHost = pluginToHostPipes.readEnd
+    self.tempDirectory = try createUniqueTemporaryDirectory(basename: "swift_wasm_macros")
+    self.stdinPath = self.tempDirectory.appending("stdin")
+    self.stdoutPath = self.tempDirectory.appending("stdout")
+
+    let stdinResult = self.stdinPath.withCString {
+      mkfifo($0, S_IRUSR | S_IWUSR)
```

Comment:
Why we need named pipe here? Wasn't `pipe(2)` enough for us?

Reply:
As mentioned in the commit, this makes the API more general and easier to test.

Reply:
I've cleaned it up in a follow-up commit (https://github.com/swiftlang/swift/pull/73031/commits/0da4f22081333a028eef181e70e1496c6ae89916)